<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>fastlmm API Documentation &mdash; fastlmm 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="fastlmm 0.0.1 documentation" href="#" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">fastlmm 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="fastlmm-api-documentation">
<h1><code class="xref py py-mod docutils literal"><span class="pre">fastlmm</span></code> API Documentation<a class="headerlink" href="#fastlmm-api-documentation" title="Permalink to this headline">¶</a></h1>
<p>FaST-LMM, which stands for Factored Spectrally Transformed Linear Mixed Models, is a program for performing both
single-SNP and SNP-set genome-wide association studies (GWAS) on extremely large data sets.
This release contains the improvements described in Widmer <em>et al.</em>, <em>Scientific Reports</em> 2014, and tests for epistasis.</p>
<p>The FaST-LMM GitHub site:
<a class="reference external" href="https://github.com/MicrosoftGenomics/FaST-LMM/">https://github.com/MicrosoftGenomics/FaST-LMM/</a></p>
<p>Our main documentation (including live examples) is also available as ipython notebook:
<a class="reference external" href="https://github.com/MicrosoftGenomics/FaST-LMM/blob/master/doc/ipynb/FaST-LMM.ipynb">https://github.com/MicrosoftGenomics/FaST-LMM/blob/master/doc/ipynb/FaST-LMM.ipynb</a></p>
<div class="section" id="single-snp">
<h2><code class="xref py py-mod docutils literal"><span class="pre">single_snp</span></code><a class="headerlink" href="#single-snp" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.single_snp">
<code class="descclassname">fastlmm.association.</code><code class="descname">single_snp</code><span class="sig-paren">(</span><em>test_snps</em>, <em>pheno</em>, <em>K0=None</em>, <em>K1=None</em>, <em>mixing=None</em>, <em>covar=None</em>, <em>covar_by_chrom=None</em>, <em>leave_out_one_chrom=True</em>, <em>output_file_name=None</em>, <em>h2=None</em>, <em>log_delta=None</em>, <em>cache_file=None</em>, <em>GB_goal=None</em>, <em>interact_with_snp=None</em>, <em>force_full_rank=False</em>, <em>force_low_rank=False</em>, <em>G0=None</em>, <em>G1=None</em>, <em>runner=None</em>, <em>count_A1=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.single_snp" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing single SNP GWAS using cross validation over the chromosomes and REML. Will reorder and intersect IIDs as needed.
(For backwards compatibility, you may use &#8216;leave_out_one_chrom=False&#8217; to skip cross validation, but that is not recommended.)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs to test. Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>pheno</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; A single phenotype: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
Any IIDs with missing values will be removed.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>K0</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string (or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>)) &#8211; SNPs from which to create a similarity matrix. If not given, will use test_snps.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(When leave_out_one_chrom is False, can be a <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code> or a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code>-formated file name.)</li>
<li><strong>K1</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string (or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>)) &#8211; SNPs from which to create a second similarity matrix, optional. (Also, see &#8216;mixing&#8217;).
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(When leave_out_one_chrom is False, can be a <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code> or a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code>-formated file name.)</li>
<li><strong>mixing</strong> (<em>number</em>) &#8211; Weight between 0.0 (inclusive, default) and 1.0 (inclusive) given to K1 relative to K0.
If you give no mixing number and a K1 is given, the best weight will be learned.</li>
<li><strong>covar</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; covariate information, optional: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>leave_out_one_chrom</strong> (<em>boolean</em>) &#8211; Perform single SNP GWAS via cross validation over the chromosomes. Default to True.
(Warning: setting False can cause proximal contamination.)</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created. The output format is tab-deleted text.</li>
<li><strong>h2</strong> (<em>number</em>) &#8211; A parameter to LMM learning, optional
If not given will search for best value.
If mixing is unspecified, then h2 must also be unspecified.</li>
<li><strong>log_delta</strong> (<em>number</em>) &#8211; a re-parameterization of h2 provided for backwards compatibility. h2 is 1./(exp(log_delta)+1)</li>
<li><strong>cache_file</strong> (<em>file name</em>) &#8211; Name of  file to read or write cached precomputation values to, optional.
If not given, no cache file will be used.
If given and file does not exist, will write precomputation values to file.
If given and file does exist, will read precomputation values from file.
The file contains the U and S matrix from the decomposition of the training matrix. It is in Python&#8217;s np.savez (*.npz) format.
Calls using the same cache file should have the same &#8216;K0&#8217; and &#8216;K1&#8217;
If given and the file does exist then K0 and K1 need not be given.</li>
<li><strong>GB_goal</strong> (<em>number</em>) &#8211; gigabytes of memory the run should use, optional. If not given, will read the test_snps in blocks the same size as the kernel,
which is memory efficient with little overhead on computation time.</li>
<li><strong>interact_with_snp</strong> &#8211; index of a covariate to perform an interaction test with. 
Allows for interaction testing (interact_with_snp x snp will be tested)
default: None</li>
<li><strong>force_full_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer SNPs than IIDs, create an explicit iid_count x iid_count kernel. Cannot be True if force_low_rank is True.</li>
<li><strong>force_low_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer IIDs than SNPs, create a low-rank iid_count x sid_count kernel. Cannot be True if force_full_rank is True.</li>
<li><strong>G0</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string (or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>)) &#8211; Same as K0. Provided for backwards compatibility. Cannot be given if K0 is given.</li>
<li><strong>G1</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string (or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>)) &#8211; Same as K1. Provided for backwards compatibility. Cannot be given if K1 is given.</li>
<li><strong>runner</strong> (<em>a runner.</em>) &#8211; a runner, optional: Tells how to run locally, multi-processor, or on a cluster.
If not given, the function is run locally.</li>
<li><strong>count_A1</strong> (<em>bool</em>) &#8211; If it needs to read SNP data from a BED-formatted file, tells if it should count the number of A1
alleles (the PLINK standard) or the number of A2 alleles. False is the current default, but in the future the default will change to True.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per test SNP. Columns include &#8220;PValue&#8221;</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">single_snp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.phe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results_dataframe</span> <span class="o">=</span> <span class="n">single_snp</span><span class="p">(</span><span class="n">test_snps</span><span class="o">=</span><span class="s2">&quot;../feature_selection/examples/toydata.5chrom&quot;</span><span class="p">,</span> <span class="n">pheno</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span><span class="nb">round</span><span class="p">(</span><span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">PValue</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">results_dataframe</span><span class="p">)</span>
<span class="go">null_576 1e-07 10000</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="single-snp-all-plus-select">
<h2><code class="xref py py-mod docutils literal"><span class="pre">single_snp_all_plus_select</span></code><a class="headerlink" href="#single-snp-all-plus-select" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.single_snp_all_plus_select">
<code class="descclassname">fastlmm.association.</code><code class="descname">single_snp_all_plus_select</code><span class="sig-paren">(</span><em>test_snps</em>, <em>pheno</em>, <em>G=None</em>, <em>covar=None</em>, <em>k_list=None</em>, <em>n_folds=10</em>, <em>seed=0</em>, <em>output_file_name=None</em>, <em>GB_goal=None</em>, <em>force_full_rank=False</em>, <em>force_low_rank=False</em>, <em>mixing=None</em>, <em>h2=None</em>, <em>do_plot=False</em>, <em>runner=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.single_snp_all_plus_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing single SNP GWAS based on two kernels. The first kernel is based on all SNPs. The second kernel is a similarity matrix
constructed of the top <em>k</em> SNPs where the SNPs are ordered via the PValue from <code class="xref py py-meth docutils literal"><span class="pre">single_snp()</span></code> and <em>k</em> is determined via out-of-sample prediction.
All work is done via &#8216;leave_out_one_chrom&#8217;, that one chromosome is tested and the kernels are constructed from the other chromosomes.
Will reorder and intersect IIDs as needed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs to test. Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>pheno</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; A single phenotype: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
Any IIDs with missing values will be removed.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>G</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs from which to create a similarity matrix of the top <em>k</em> SNPs. If not given, will use test_snps.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>covar</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; covariate information, optional: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>k_list</strong> (<em>list of numbers</em>) &#8211; Values of <em>k</em> (in addition to 0) to test. Default to [1,2,4,8,...8192].</li>
<li><strong>n_folds</strong> (<em>number</em>) &#8211; Number of folds of cross validation to use for out-of-sample evaluation of various values of <em>k</em>. Default to 10.</li>
<li><strong>seed</strong> (<em>number</em>) &#8211; (optional) Random seed used to generate permutations for lrt G0 fitting.</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created.</li>
<li><strong>GB_goal</strong> (<em>number</em>) &#8211; gigabytes of memory the run should use, optional. If not given, will read the test_snps in blocks the same size as the kernel,
which is memory efficient with little overhead on computation time.</li>
<li><strong>force_full_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer SNPs than IIDs, create an explicit iid_count x iid_count kernel. Cannot be True if force_low_rank is True.</li>
<li><strong>force_low_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer IIDs than SNPs, create a low-rank iid_count x sid_count kernel. Cannot be True if force_full_rank is True.</li>
<li><strong>mixing</strong> (<em>number</em>) &#8211; A parameter to LMM learning telling how to combine the two kernels, optional
If not given will search for best value.</li>
<li><strong>h2</strong> (<em>number</em>) &#8211; A parameter to LMM learning that tells how much weight to give the K&#8217;s vs. the identity matrix, optional
If not given will search for best value.</li>
<li><strong>do_plot</strong> (<em>boolean</em>) &#8211; If true, will plot, for each chrom, the negative loglikelihood vs k.</li>
<li><strong>runner</strong> (<em>a runner.</em>) &#8211; a runner, optional: Tells how to run locally, multi-processor, or on a cluster.
If not given, the function is run locally.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per test SNP. Columns include &#8220;PValue&#8221;</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">single_snp_all_plus_select</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.util.runner</span> <span class="k">import</span> <span class="n">LocalMultiProc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.phe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">snps</span> <span class="o">=</span> <span class="n">Bed</span><span class="p">(</span><span class="s2">&quot;../feature_selection/examples/toydata.5chrom.bed&quot;</span><span class="p">)[:,::</span><span class="mi">100</span><span class="p">]</span> <span class="c1">#To make example faster, run on only 1/100th of the data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chrom5_snps</span> <span class="o">=</span> <span class="n">snps</span><span class="p">[:,</span><span class="n">snps</span><span class="o">.</span><span class="n">pos</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="mi">5</span><span class="p">]</span> <span class="c1"># Test on only chrom5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results_dataframe</span> <span class="o">=</span> <span class="n">single_snp_all_plus_select</span><span class="p">(</span><span class="n">test_snps</span><span class="o">=</span><span class="n">chrom5_snps</span><span class="p">,</span><span class="n">G</span><span class="o">=</span><span class="n">snps</span><span class="p">,</span><span class="n">pheno</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">,</span><span class="n">GB_goal</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">runner</span><span class="o">=</span><span class="n">LocalMultiProc</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="n">mkl_num_threads</span><span class="o">=</span><span class="mi">5</span><span class="p">))</span> <span class="c1">#Run multiproc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span><span class="nb">round</span><span class="p">(</span><span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">PValue</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">results_dataframe</span><span class="p">)</span>
<span class="go">null_9800 0.0793385 4</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="single-snp-linreg">
<h2><code class="xref py py-mod docutils literal"><span class="pre">single_snp_linreg</span></code><a class="headerlink" href="#single-snp-linreg" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.single_snp_linreg">
<code class="descclassname">fastlmm.association.</code><code class="descname">single_snp_linreg</code><span class="sig-paren">(</span><em>test_snps</em>, <em>pheno</em>, <em>covar=None</em>, <em>max_output_len=None</em>, <em>output_file_name=None</em>, <em>GB_goal=None</em>, <em>runner=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.single_snp_linreg" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing single SNP GWAS using linear regression. Will reorder and intersect IIDs as needed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs to test. Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>pheno</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; A single phenotype: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
Any IIDs with missing values will be removed.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>covar</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; covariate information, optional: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>max_output_len</strong> (<em>number</em>) &#8211; Maximum number of Pvalues to return. Default to None, which means &#8216;Return all&#8217;.</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created. The output format is tab-deleted text.</li>
<li><strong>GB_goal</strong> (<em>number</em>) &#8211; gigabytes of memory the run should use, optional. If not given, will read the test_snps in blocks of size iid_count,
which is memory efficient with little overhead on computation time.</li>
<li><strong>runner</strong> (<em>a runner.</em>) &#8211; a runner, optional: Tells how to run locally, multi-processor, or on a cluster.
If not given, the function is run locally.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per test SNP. Columns include &#8220;PValue&#8221;</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">single_snp_linreg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.phe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results_dataframe</span> <span class="o">=</span> <span class="n">single_snp_linreg</span><span class="p">(</span><span class="n">test_snps</span><span class="o">=</span><span class="s2">&quot;../feature_selection/examples/toydata.5chrom&quot;</span><span class="p">,</span> <span class="n">pheno</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span><span class="nb">round</span><span class="p">(</span><span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">PValue</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">results_dataframe</span><span class="p">)</span>
<span class="go">null_576 1e-07 10000</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="single-snp-select">
<h2><code class="xref py py-mod docutils literal"><span class="pre">single_snp_select</span></code><a class="headerlink" href="#single-snp-select" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.single_snp_select">
<code class="descclassname">fastlmm.association.</code><code class="descname">single_snp_select</code><span class="sig-paren">(</span><em>test_snps</em>, <em>pheno</em>, <em>G=None</em>, <em>covar=None</em>, <em>k_list=None</em>, <em>n_folds=10</em>, <em>just_return_selected_snps=False</em>, <em>seed=0</em>, <em>output_file_name=None</em>, <em>GB_goal=None</em>, <em>force_full_rank=False</em>, <em>force_low_rank=False</em>, <em>h2=None</em>, <em>runner=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.single_snp_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing single SNP GWAS based on covariates (often PCs) and a similarity matrix constructed of the top <em>k</em> SNPs where
SNPs are ordered via the PValue from <code class="xref py py-meth docutils literal"><span class="pre">single_snp_linreg()</span></code> and <em>k</em> is determined via out-of-sample prediction. Will reorder and intersect IIDs as needed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs to test. Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>pheno</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; A single phenotype: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
Any IIDs with missing values will be removed.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>G</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs from which to create a similarity matrix of the top <em>k</em> SNPs. If not given, will use test_snps.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>covar</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; covariate information, optional: Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>, for example, <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.
(For backwards compatibility can also be dictionary with keys &#8216;vals&#8217;, &#8216;iid&#8217;, &#8216;header&#8217;)</li>
<li><strong>k_list</strong> (<em>list of numbers</em>) &#8211; Values of <em>k</em> (in addition to 0) to test. Default to [1,2,4,8,...8192].</li>
<li><strong>n_folds</strong> (<em>number</em>) &#8211; Number of folds of cross validation to use for out-of-sample evaluation of various values of <em>k</em>. Default to 10.</li>
<li><strong>just_return_selected_snps</strong> (<em>list of strings</em>) &#8211; Instead of returning the results of GWAS, return the top <em>k</em> SNPs selected.</li>
<li><strong>seed</strong> (<em>number</em>) &#8211; (optional) Random seed used to generate permutations for lrt G0 fitting.</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created.</li>
<li><strong>GB_goal</strong> (<em>number</em>) &#8211; gigabytes of memory the run should use, optional. If not given, will read the test_snps in blocks the same size as the kernel,
which is memory efficient with little overhead on computation time.</li>
<li><strong>force_full_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer SNPs than IIDs, create an explicit iid_count x iid_count kernel. Cannot be True if force_low_rank is True.</li>
<li><strong>force_low_rank</strong> (<em>Boolean</em>) &#8211; Even if kernels are defined with fewer IIDs than SNPs, create a low-rank iid_count x sid_count kernel. Cannot be True if force_full_rank is True.</li>
<li><strong>h2</strong> (<em>number</em>) &#8211; A parameter to LMM learning that tells how much weight to give the K&#8217;s vs. the identity matrix, optional
If not given will search for best value.</li>
<li><strong>runner</strong> (<em>a runner.</em>) &#8211; a runner, optional: Tells how to run locally, multi-processor, or on a cluster.
If not given, the function is run locally.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per test SNP. Columns include &#8220;PValue&#8221;</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">single_snp_select</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.util</span> <span class="k">import</span> <span class="n">compute_auto_pcs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_fn</span> <span class="o">=</span> <span class="s2">&quot;../../tests/datasets/synth/all.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phen_fn</span> <span class="o">=</span> <span class="s2">&quot;../../tests/datasets/synth/pheno_10_causals.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">covar</span> <span class="o">=</span> <span class="n">compute_auto_pcs</span><span class="p">(</span><span class="n">bed_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results_dataframe</span> <span class="o">=</span> <span class="n">single_snp_select</span><span class="p">(</span><span class="n">test_snps</span><span class="o">=</span><span class="n">bed_fn</span><span class="p">,</span> <span class="n">G</span><span class="o">=</span><span class="n">bed_fn</span><span class="p">,</span> <span class="n">pheno</span><span class="o">=</span><span class="n">phen_fn</span><span class="p">,</span> <span class="n">covar</span><span class="o">=</span><span class="n">covar</span><span class="p">,</span> <span class="n">GB_goal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span><span class="nb">round</span><span class="p">(</span><span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">PValue</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">results_dataframe</span><span class="p">)</span>
<span class="go">snp495_m0_.01m1_.04 0.0 5000</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="epistasis">
<h2><code class="xref py py-mod docutils literal"><span class="pre">epistasis</span></code><a class="headerlink" href="#epistasis" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.epistasis">
<code class="descclassname">fastlmm.association.</code><code class="descname">epistasis</code><span class="sig-paren">(</span><em>test_snps</em>, <em>pheno</em>, <em>G0</em>, <em>G1=None</em>, <em>mixing=0.0</em>, <em>covar=None</em>, <em>output_file_name=None</em>, <em>sid_list_0=None</em>, <em>sid_list_1=None</em>, <em>log_delta=None</em>, <em>min_log_delta=-5</em>, <em>max_log_delta=10</em>, <em>cache_file=None</em>, <em>runner=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.epistasis" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing epistasis GWAS with ML (never REML).  See <a class="reference external" href="http://www.nature.com/srep/2013/130122/srep01099/full/srep01099.html">http://www.nature.com/srep/2013/130122/srep01099/full/srep01099.html</a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs from which to test pairs. If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>pheno</strong> (<em>a 'pheno dictionary' or a string</em>) &#8211; A single phenotype: A &#8216;pheno dictionary&#8217; contains an ndarray on the &#8216;vals&#8217; key and a iid list on the &#8216;iid&#8217; key.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>G0</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs from which to construct a similarity matrix.
If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>G1</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs from which to construct a second similarity kernel, optional. Also, see &#8216;mixing&#8217;).
If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>mixing</strong> (<em>number</em>) &#8211; Weight between 0.0 (inclusive, default) and 1.1 (inclusive) given to G1 relative to G0.
If you give no mixing number, G0 will get all the weight and G1 will be ignored.</li>
<li><strong>covar</strong> (<em>a 'pheno dictionary' or a string</em>) &#8211; covariate information, optional: A &#8216;pheno dictionary&#8217; contains an ndarray on the &#8216;vals&#8217; key and a iid list on the &#8216;iid&#8217; key.
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>sid_list_0</strong> (<em>list of strings</em>) &#8211; list of sids, optional:
All unique pairs from sid_list_0 x sid_list_1 will be evaluated.
If you give no sid_list_0, all sids in test_snps will be used.</li>
<li><strong>sid_list_1</strong> (<em>list of strings</em>) &#8211; list of sids, optional:
All unique pairs from sid_list_0 x sid_list_1 will be evaluated.
If you give no sid_list_1, all sids in test_snps will be used.</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created. The output format is tab-deleted text.</li>
<li><strong>log_delta</strong> (<em>number</em>) &#8211; A parameter to LMM learning, optional
If not given will search for best value.</li>
<li><strong>min_log_delta</strong> (<em>number</em>) &#8211; (default:-5)
When searching for log_delta, the lower bounds of the search.</li>
<li><strong>max_log_delta</strong> (<em>number</em>) &#8211; (default:-5)
When searching for log_delta, the upper bounds of the search.</li>
<li><strong>cache_file</strong> (<em>file name</em>) &#8211; Name of  file to read or write cached precomputation values to, optional.
If not given, no cache file will be used.
If given and file does not exists, will write precomputation values to file.
If given and file does exists, will read precomputation values from file.
The file contains the U and S matrix from the decomposition of the training matrix. It is in Python&#8217;s np.savez (*.npz) format.
Calls using the same cache file should have the same &#8216;G0&#8217; and &#8216;G1&#8217;</li>
<li><strong>runner</strong> (<em>a runner.</em>) &#8211; a runner, optional: Tells how to run locally, multi-processor, or on a cluster.
If not given, the function is run locally.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per SNP pair. Columns include &#8220;PValue&#8221;</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">epistasis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test_snps</span> <span class="o">=</span> <span class="n">Bed</span><span class="p">(</span><span class="s1">&#39;../../tests/datasets/all_chr.maf0.001.N300&#39;</span><span class="p">,</span><span class="n">count_A1</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno</span> <span class="o">=</span> <span class="s1">&#39;../../tests/datasets/phenSynthFrom22.23.N300.randcidorder.txt&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">covar</span> <span class="o">=</span> <span class="s1">&#39;../../tests/datasets/all_chr.maf0.001.covariates.N300.txt&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results_dataframe</span> <span class="o">=</span> <span class="n">epistasis</span><span class="p">(</span><span class="n">test_snps</span><span class="p">,</span> <span class="n">pheno</span><span class="p">,</span> <span class="n">G0</span><span class="o">=</span><span class="n">test_snps</span><span class="p">,</span> <span class="n">covar</span><span class="o">=</span><span class="n">covar</span><span class="p">,</span> 
<span class="gp">... </span>                                <span class="n">sid_list_0</span><span class="o">=</span><span class="n">test_snps</span><span class="o">.</span><span class="n">sid</span><span class="p">[:</span><span class="mi">10</span><span class="p">],</span> <span class="c1">#first 10 snps</span>
<span class="gp">... </span>                                <span class="n">sid_list_1</span><span class="o">=</span><span class="n">test_snps</span><span class="o">.</span><span class="n">sid</span><span class="p">[</span><span class="mi">5</span><span class="p">:</span><span class="mi">15</span><span class="p">])</span> <span class="c1">#Skip 5 snps, use next 10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP0</span><span class="p">,</span> <span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SNP1</span><span class="p">,</span><span class="nb">round</span><span class="p">(</span><span class="n">results_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">PValue</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">results_dataframe</span><span class="p">)</span>
<span class="go">1_12 1_9 0.07779 85</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="snp-set">
<h2><code class="xref py py-mod docutils literal"><span class="pre">snp_set</span></code><a class="headerlink" href="#snp-set" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.snp_set">
<code class="descclassname">fastlmm.association.</code><code class="descname">snp_set</code><span class="sig-paren">(</span><em>test_snps</em>, <em>set_list</em>, <em>pheno</em>, <em>covar=None</em>, <em>output_file_name=None</em>, <em>G0=None</em>, <em>test='lrt'</em>, <em>write_lrtperm=False</em>, <em>nperm=10</em>, <em>npermabs=None</em>, <em>mpheno=1</em>, <em>G0_fit='qq'</em>, <em>qmax=0.1</em>, <em>seed=None</em>, <em>minsetsize=None</em>, <em>maxsetsize=None</em>, <em>mindist=0</em>, <em>idist=1</em>, <em>show_pvalue_5050=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.snp_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Function performing GWAS on sets of snps</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>test_snps</strong> (<em>a string</em>) &#8211; The base name of the file containing the SNPs for alternative kernel. The file must be in PLINK Bed format.</li>
<li><strong>set_list</strong> (<em>a string</em>) &#8211; The name of a tab-delimited file defining the sets. The file should contain two-columns &#8216;snp&#8217; and &#8216;set&#8217;.</li>
<li><strong>pheno</strong> (<em>a string</em>) &#8211; The name of a file containing the phenotype. The file must be in PLINK phenotype format.</li>
<li><strong>covar</strong> (<em>a 'pheno dictionary' or a string</em>) &#8211; covariate information, optional: The name of a file in PLINK phenotype format.</li>
<li><strong>output_file_name</strong> (<em>file name</em>) &#8211; Name of file to write results to, optional. If not given, no output file will be created.</li>
<li><strong>G0</strong> (<em>a string</em>) &#8211; Training SNPs from which to construct a similarity kernel. It should be the base name of files in PLINK Bed or Ped format.</li>
<li><strong>test</strong> (<em>a string</em>) &#8211; &#8216;lrt&#8217; (default) or &#8216;sc_davies&#8217;</li>
<li><strong>write_lrtperm</strong> (<em>boolean</em>) &#8211; (default: False) If True, write the lrtperm vector (dictated by seed) to a file.</li>
<li><strong>nperm</strong> (<em>number</em>) &#8211; (default: 10) number of permutations per test</li>
<li><strong>npermabs</strong> (<em>number</em>) &#8211; (default: None) absolute number of permutations</li>
<li><strong>mpheno</strong> (<em>number</em>) &#8211; (default: 1) integer, starting at 1, representing the index of the phenotype tested</li>
<li><strong>G0_fit</strong> (<em>string</em>) &#8211; (default: &#8220;qq&#8221;) How to fit G0. Should be either &#8220;qq&#8221; or &#8220;ml&#8221;</li>
<li><strong>qmax</strong> (<em>number</em>) &#8211; (default: .1) Use the top qmax fraction of G0 distrib test statistics to fit the G0 distribution</li>
<li><strong>seed</strong> (<em>number</em>) &#8211; (optional) Random seed used to generate permutations for lrt G0 fitting.</li>
<li><strong>minsetsize</strong> (<em>number</em>) &#8211; (optional) only include sets at least this large (inclusive)</li>
<li><strong>maxsetsize</strong> (<em>number</em>) &#8211; (optional) only include sets no more than this large (inclusive)</li>
<li><strong>mindist</strong> (<em>number</em>) &#8211; (default 0) SNPs within mindist from the test SNPs will be removed from</li>
<li><strong>idist</strong> (<em>number</em>) &#8211; (default: 1) the type of position to use with mindist
1, genomic distance
2, base-pair distance</li>
<li><strong>show_pvalue_5050</strong> (<em>Boolean</em>) &#8211; (default: False) show a conservative P-value arising from an assumed null distribution that is a 50-50 mixture distribution of 0 and 1 dof chi squares [Molenberghs and Verbeke, 2003].
Provided for backwards compatibility.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">Pandas dataframe with one row per set.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.association</span> <span class="k">import</span> <span class="n">snp_set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result_dataframe</span> <span class="o">=</span> <span class="n">snp_set</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">test_snps</span> <span class="o">=</span> <span class="s1">&#39;../../tests/datasets/all_chr.maf0.001.N300&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">set_list</span> <span class="o">=</span> <span class="s1">&#39;../../tests/datasets/set_input.23.txt&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">pheno</span> <span class="o">=</span> <span class="s1">&#39;../../tests/datasets/phenSynthFrom22.23.N300.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">result_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">SetId</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">result_dataframe</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;P-value&#39;</span><span class="p">],</span><span class="mi">15</span><span class="p">)</span>
<span class="go">set23 0.0</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="fastlmm">
<h2><code class="xref py py-class docutils literal"><span class="pre">FastLMM</span></code><a class="headerlink" href="#fastlmm" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="fastlmm.inference.FastLMM">
<em class="property">class </em><code class="descclassname">fastlmm.inference.</code><code class="descname">FastLMM</code><span class="sig-paren">(</span><em>GB_goal=None</em>, <em>force_full_rank=False</em>, <em>force_low_rank=False</em>, <em>snp_standardizer=Unit()</em>, <em>covariate_standardizer=Unit()</em>, <em>kernel_standardizer=DiagKtoN()</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.FastLMM" title="Permalink to this definition">¶</a></dt>
<dd><p>A predictor, somewhat in the style of scikit-learn, for learning and predicting with linear mixed models.</p>
<dl class="docutils">
<dt><strong>Constructor:</strong></dt>
<dd><table class="first docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>GB_goal</strong> (int) &#8211; gigabytes of memory the run should use, optional. If not given, will read the test_snps in blocks the same size as the kernel, which is memory efficient with little overhead on computation time.</li>
<li><strong>force_full_rank</strong> (bool) &#8211; Even if kernels are defined with fewer SNPs than IIDs, create an explicit iid_count x iid_count kernel. Cannot be True if force_low_rank is True.</li>
<li><strong>force_low_rank</strong> (bool) &#8211; Even if kernels are defined with fewer IIDs than SNPs, create a low-rank iid_count x sid_count kernel. Cannot be True if force_full_rank is True.</li>
<li><strong>snp_standardizer</strong> (<code class="xref py py-class docutils literal"><span class="pre">Standardizer</span></code>) &#8211; The PySnpTools standardizer to be apply to SNP data. Choices include <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Unit</span></code> (Default. Makes values for each SNP have mean zero and standard deviation 1.0, then fills missing with zero) and <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Identity</span></code> (Do nothing)</li>
<li><strong>covariate_standardizer</strong> (<code class="xref py py-class docutils literal"><span class="pre">Standardizer</span></code>) &#8211; The PySnpTools standardizer to be apply to X, the covariate data. Some choices include <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Unit</span></code> (Default. Fills missing with zero) and <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Identity</span></code> (do nothing)</li>
<li><strong>kernel_standardizer</strong> (<code class="xref py py-class docutils literal"><span class="pre">KernelStandardizer</span></code>) &#8211; The PySnpTools kernel standardizer to be apply to the kernels. Some choices include <code class="xref py py-class docutils literal"><span class="pre">KernelStandardizer.DiagKToN</span></code> (Default. Make the diagonal sum to iid_count)  and <code class="xref py py-class docutils literal"><span class="pre">KernelStandardizer.Identity</span></code> (Do nothing)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="last highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Bed</span><span class="p">,</span> <span class="n">Pheno</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.inference</span> <span class="k">import</span> <span class="n">FastLMM</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">snpreader</span> <span class="o">=</span> <span class="n">Bed</span><span class="p">(</span><span class="s1">&#39;../feature_selection/examples/toydata.bed&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cov_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.cov&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.phe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">train_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="n">snpreader</span><span class="o">.</span><span class="n">iid_count</span><span class="p">]</span> <span class="c1"># iids 10 and on</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test_idx</span>  <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span> <span class="c1"># the first 10 iids</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fastlmm</span> <span class="o">=</span> <span class="n">FastLMM</span><span class="p">(</span><span class="n">GB_goal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#We give it phenotype and covariate information for extra examples, but it reorders and intersects the examples, so only training examples are used. </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">fastlmm</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">K0_train</span><span class="o">=</span><span class="n">snpreader</span><span class="p">[</span><span class="n">train_idx</span><span class="p">,:],</span><span class="n">X</span><span class="o">=</span><span class="n">cov_fn</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">mean</span><span class="p">,</span> <span class="n">covariance</span> <span class="o">=</span> <span class="n">fastlmm</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">K0_whole_test</span><span class="o">=</span><span class="n">snpreader</span><span class="p">[</span><span class="n">test_idx</span><span class="p">,:],</span><span class="n">X</span><span class="o">=</span><span class="n">cov_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">mean</span><span class="o">.</span><span class="n">iid</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean</span><span class="o">.</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">7</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">covariance</span><span class="o">.</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="mi">7</span><span class="p">)</span>
<span class="go">[&#39;per0&#39; &#39;per0&#39;] 0.1791958 0.8995209</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nll</span> <span class="o">=</span> <span class="n">fastlmm</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">K0_whole_test</span><span class="o">=</span><span class="n">snpreader</span><span class="p">[</span><span class="n">test_idx</span><span class="p">,:],</span><span class="n">X</span><span class="o">=</span><span class="n">cov_fn</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">round</span><span class="p">(</span><span class="n">nll</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
<span class="go">13.4623234</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="method">
<dt id="fastlmm.inference.FastLMM.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>X=None</em>, <em>y=None</em>, <em>K0_train=None</em>, <em>K1_train=None</em>, <em>h2raw=None</em>, <em>mixing=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.FastLMM.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for training a <code class="xref py py-class docutils literal"><span class="pre">FastLMM</span></code> predictor. If the examples in X, y, K0_train, K1_train are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; training covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>y</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; training phenotype:
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_train</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A similarity matrix or SNPs from which to construct such a similarity matrix.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>K1_train</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A second similarity matrix or SNPs from which to construct such a second similarity matrix. (Also, see &#8216;mixing&#8217;).
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>h2raw</strong> (<em>number</em>) &#8211; A parameter to LMM learning that tells how much weight to give the K&#8217;s vs. the identity matrix, optional 
If not given will search for best value.
If mixing is unspecified, then h2 must also be unspecified.</li>
<li><strong>mixing</strong> (<em>number</em>) &#8211; Weight between 0.0 (inclusive, default) and 1.0 (inclusive) given to K1_train relative to K0_train.
If you give no mixing number and a K1_train is given, the best weight will be learned.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">self, the fitted FastLMM predictor</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="fastlmm.inference.FastLMM.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>X=None</em>, <em>K0_whole_test=None</em>, <em>K1_whole_test=None</em>, <em>iid_if_none=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.FastLMM.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for predicting from a fitted <code class="xref py py-class docutils literal"><span class="pre">FastLMM</span></code> predictor.
If the examples in X, K0_whole_test, K1_whole_test are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A similarity matrix from all the examples to the test examples. Alternatively,
the test SNPs needed to construct such a similarity matrix.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>K1_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A second similarity matrix from all the examples to the test examples. Alternatively,
the test SNPs needed to construct such a similarity matrix.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>iid_if_none</strong> (<em>an ndarray of two strings</em>) &#8211; Examples to predict for if no X, K0_whole_test, K1_whole_test is provided.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">A <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code> of the means and a <code class="xref py py-class docutils literal"><span class="pre">KernelData</span></code> of the covariance</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="fastlmm.inference.FastLMM.score">
<code class="descname">score</code><span class="sig-paren">(</span><em>X=None</em>, <em>y=None</em>, <em>K0_whole_test=None</em>, <em>K1_whole_test=None</em>, <em>iid_if_none=None</em>, <em>return_mse_too=False</em>, <em>return_per_iid=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.FastLMM.score" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating the negative log likelihood of testing examples.
If the examples in X,y,  K0_whole_test, K1_whole_test are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>y</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing phenotype:
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A similarity matrix from all the examples to the test examples. Alternatively,
the test SNPs needed to construct such a similarity matrix.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>K1_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; A second similarity matrix from all the examples to the test examples. Alternatively,
the test SNPs needed to construct such a similarity matrix.
Can be any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>. If you give a string, can be the name of a PLINK-formated Bed file.
Can be PySnpTools <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>. If you give a string it can be the name of a <code class="xref py py-class docutils literal"><span class="pre">KernelNpz</span></code> file.</li>
<li><strong>iid_if_none</strong> (<em>an ndarray of two strings</em>) &#8211; Examples to predict for if no X, K0_whole_test, K1_whole_test is provided.</li>
<li><strong>return_mse_too</strong> (<em>bool</em>) &#8211; If true, will also return the mean squared error.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a float of the negative log likelihood and, optionally, a float of the mean squared error.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="linearregression">
<h2><code class="xref py py-class docutils literal"><span class="pre">LinearRegression</span></code><a class="headerlink" href="#linearregression" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="fastlmm.inference.LinearRegression">
<em class="property">class </em><code class="descclassname">fastlmm.inference.</code><code class="descname">LinearRegression</code><span class="sig-paren">(</span><em>covariate_standardizer=Unit()</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.LinearRegression" title="Permalink to this definition">¶</a></dt>
<dd><p>A linear regression predictor, that works like the FastLMM in fastlmm_predictor.py, but that expects all similarity matrices to be identity.</p>
<dl class="docutils">
<dt><strong>Constructor:</strong></dt>
<dd><table class="first docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>covariate_standardizer</strong> (<code class="xref py py-class docutils literal"><span class="pre">Standardizer</span></code>) &#8211; The PySnpTools standardizer to be apply to X, the covariate data. Some choices include <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Unit</span></code> (Default. Fills missing with zero) and <code class="xref py py-class docutils literal"><span class="pre">Standardizer.Identity</span></code> (do nothing)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="last highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="k">import</span> <span class="n">Pheno</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.inference</span> <span class="k">import</span> <span class="n">LinearRegression</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cov</span> <span class="o">=</span> <span class="n">Pheno</span><span class="p">(</span><span class="s2">&quot;../feature_selection/examples/toydata.cov&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pheno_fn</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata.phe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">train_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="n">cov</span><span class="o">.</span><span class="n">iid_count</span><span class="p">]</span> <span class="c1"># iids 10 and on</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test_idx</span>  <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span> <span class="c1"># the first 10 iids</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">linreg</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#We give it phenotype information for extra examples, but it reorders and intersects the examples, so only training examples are used. </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">linreg</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">cov</span><span class="p">[</span><span class="n">train_idx</span><span class="p">,:],</span><span class="n">y</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">mean</span><span class="p">,</span> <span class="n">covariance</span> <span class="o">=</span> <span class="n">linreg</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">cov</span><span class="p">[</span><span class="n">test_idx</span><span class="p">,:])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">mean</span><span class="o">.</span><span class="n">iid</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean</span><span class="o">.</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">7</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">covariance</span><span class="o">.</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="mi">7</span><span class="p">)</span>
<span class="go">[&#39;per0&#39; &#39;per0&#39;] 0.1518764 0.9043703</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nll</span> <span class="o">=</span> <span class="n">linreg</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">cov</span><span class="p">[</span><span class="n">test_idx</span><span class="p">,:],</span><span class="n">y</span><span class="o">=</span><span class="n">pheno_fn</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">round</span><span class="p">(</span><span class="n">nll</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
<span class="go">13.6688448</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="method">
<dt id="fastlmm.inference.LinearRegression.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>X=None</em>, <em>y=None</em>, <em>K0_train=None</em>, <em>K1_train=None</em>, <em>h2=None</em>, <em>mixing=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.LinearRegression.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for training a <code class="xref py py-class docutils literal"><span class="pre">FastLMM</span></code> predictor. If the examples in X, y, K0_train, K1_train are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; training covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>y</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; training phenotype:
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_train</strong> (<em>None</em>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>K1_train</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>h2</strong> (<em>number</em>) &#8211; Ignored. Optional.</li>
<li><strong>mixing</strong> (<em>number</em>) &#8211; Ignored. Optional.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">self, the fitted Linear Regression predictor</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="fastlmm.inference.LinearRegression.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>X=None</em>, <em>K0_whole_test=None</em>, <em>K1_whole_test=None</em>, <em>iid_if_none=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.LinearRegression.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for predicting from a fitted <code class="xref py py-class docutils literal"><span class="pre">FastLMM</span></code> predictor.
If the examples in X, K0_whole_test, K1_whole_test are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_whole_test</strong> (<em>None</em>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>K1_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>iid_if_none</strong> (<em>an ndarray of two strings</em>) &#8211; Examples to predict for if no X, K0_whole_test, K1_whole_test is provided.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">A <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code> of the means and a <code class="xref py py-class docutils literal"><span class="pre">KernelData</span></code> of the covariance</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="fastlmm.inference.LinearRegression.score">
<code class="descname">score</code><span class="sig-paren">(</span><em>X=None</em>, <em>y=None</em>, <em>K0_whole_test=None</em>, <em>K1_whole_test=None</em>, <em>iid_if_none=None</em>, <em>return_mse_too=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.inference.LinearRegression.score" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating the negative log likelihood of testing examples.
If the examples in X,y,  K0_whole_test, K1_whole_test are not the same, they will be reordered and intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>X</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing covariate information, optional: 
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>y</strong> (a PySnpTools <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> (such as <code class="xref py py-class docutils literal"><span class="pre">Pheno</span></code> or <code class="xref py py-class docutils literal"><span class="pre">SnpData</span></code>) or string.) &#8211; testing phenotype:
If you give a string, it should be the file name of a PLINK phenotype-formatted file.</li>
<li><strong>K0_whole_test</strong> (<em>None</em>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>K1_whole_test</strong> (<code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string or <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>) &#8211; Must be None. Represents the identity similarity matrix.</li>
<li><strong>iid_if_none</strong> (<em>an ndarray of two strings</em>) &#8211; Examples to predict for if no X, K0_whole_test, K1_whole_test is provided.</li>
<li><strong>return_mse_too</strong> (<em>bool</em>) &#8211; If true, will also return the mean squared error.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a float of the negative log likelihood and, optionally, a float of the mean squared error.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="heritability-spatial-correction">
<h2><code class="xref py py-mod docutils literal"><span class="pre">heritability_spatial_correction</span></code><a class="headerlink" href="#heritability-spatial-correction" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.association.heritability_spatial_correction">
<code class="descclassname">fastlmm.association.</code><code class="descname">heritability_spatial_correction</code><span class="sig-paren">(</span><em>G_kernel</em>, <em>spatial_coor</em>, <em>spatial_iid</em>, <em>alpha_list</em>, <em>alpha_power</em>, <em>pheno</em>, <em>map_function=&lt;built-in function map&gt;</em>, <em>cache_folder=None</em>, <em>jackknife_count=500</em>, <em>permute_plus_count=10000</em>, <em>permute_times_count=10000</em>, <em>seed=0</em>, <em>just_testing=False</em>, <em>always_remote=False</em>, <em>allow_gxe2=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.association.heritability_spatial_correction" title="Permalink to this definition">¶</a></dt>
<dd><p>Function measuring heritability with correction for spatial location.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>G_kernel</strong> (a <code class="xref py py-class docutils literal"><span class="pre">KernelReader</span></code>, <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; A kernel that tells the genetic similarity between all pairs of individuals. The kernel can be given 
explicitly, for example with a <code class="xref py py-class docutils literal"><span class="pre">KernelData</span></code>. The kernel can also be given implicitly by providing a set of
SNPs or the name of a BED file.</li>
<li><strong>spatial_coor</strong> (<em>a iid_count x 2 array</em>) &#8211; The position of each individual given by two coordinates. Any units are allowed, but the two values
must be compatible so that distance can be determined via Pythagoras&#8217; theorem. (So, longitude and latitude should
not be used unless the locations are near the Equator.)</li>
<li><strong>spatial_iid</strong> (<em>array of strings with shape [iid_count,2]</em>) &#8211; A ndarray of the iids. Each iid is a ndarray of two strings (a family ID and a case ID) that identifies an individual.</li>
<li><strong>alpha_list</strong> (<em>number</em>) &#8211; a list of numbers to search to find the best alpha, which is the similarity scale. The similarity of two individuals
is here defined as exp(-(distance_between/alpha)**alpha_power). If the closest individuals are 100 units apart and the farthest
individuals are 4e6 units apart, a reasonable alpha_list might be: [int(v) for v in np.logspace(np.log10(100),np.log10(1e10), 100)]
The function&#8217;s reports on the alphas chosen. If an extreme alpha is picked, change alpha_list to cover more range.</li>
<li><strong>alpha_power</strong> &#8211; 2 (a good choice) means that similarity goes with area. 1 means with distance.</li>
<li><strong>pheno</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or string) &#8211; The target values(s) to predict. It can be a file name readable via <code class="xref py py-class docutils literal"><span class="pre">SnpReader.Pheno</span></code> or any <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code>.</li>
<li><strong>cache_folder</strong> (<em>a string</em>) &#8211; (default &#8216;None&#8217;) The name of a directory in which to save intermediate results. If &#8216;None&#8217;, then no intermediate results are saved.</li>
<li><strong>map_function</strong> (<em>a function</em>) &#8211; (default &#8216;map&#8217;) A function with the same inputs and functionality as Python&#8217;s &#8216;map&#8217; function.
Can be used to run &#8216;heritability_spatial_correction&#8217; on a cluster.</li>
<li><strong>jackknife_count</strong> (<em>number</em>) &#8211; (default 500) The number of jackknife groups to use when calculating standard errors (SE). Changing to a small number, 2, 
speeds up calculation at the cost of unusable SEs.</li>
<li><strong>permute_plus_count</strong> (<em>number</em>) &#8211; (default 10000) The number of permutations used when calculating P values. Changing to a small number, 1, 
speeds up calculation at the cost of unusable P values.</li>
<li><strong>permute_times_count</strong> (<em>number</em>) &#8211; (default 10000) The number of permutations used when calculating P values. Changing to a small number, 1, 
speeds up calculation at the cost of unusable P values.</li>
<li><strong>seed</strong> (<em>number</em>) &#8211; (default 0) The random seed used by jackknifing and permutation.</li>
<li><strong>just_testing</strong> (<em>bool</em>) &#8211; (default False) If true, skips actual LMM-related search and calculation.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Pandas dataframe with one row per phenotyper. Columns include &#8220;h2uncorr&#8221;, &#8220;h2corr&#8221;, etc.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="compute-auto-pcs">
<h2><code class="xref py py-mod docutils literal"><span class="pre">compute_auto_pcs</span></code><a class="headerlink" href="#compute-auto-pcs" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="fastlmm.util.compute_auto_pcs">
<code class="descclassname">fastlmm.util.</code><code class="descname">compute_auto_pcs</code><span class="sig-paren">(</span><em>snpreader</em>, <em>cutoff=0.1</em>, <em>k_values=array([ 0</em>, <em>1</em>, <em>2</em>, <em>3</em>, <em>4</em>, <em>5</em>, <em>6</em>, <em>7</em>, <em>8</em>, <em>9</em>, <em>10])</em>, <em>output_file_name=None</em>, <em>count_A1=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fastlmm.util.compute_auto_pcs" title="Permalink to this definition">¶</a></dt>
<dd><p>Function automatically finds the best principle components (PCs)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>snpreader</strong> (a <code class="xref py py-class docutils literal"><span class="pre">SnpReader</span></code> or a string) &#8211; SNPs for which to find the best PCs
If you give a string, it should be the base name of a set of PLINK Bed-formatted files.</li>
<li><strong>cutoff</strong> (<em>a number between 0 and 1.</em>) &#8211; (default: .1) The degree of relatedness to remove before finding the best number of PCs.
Relatedness is measured with a RRM (realized relationship matrix) so 0 is no relation, .5 is a sibling or parent, and 1 is self or twin.</li>
<li><strong>k_values</strong> (<em>list of integers</em>) &#8211; (default: 0 ... 10 [inclusive]) The number of PCs to search.</li>
<li><strong>count_A1</strong> (<em>bool</em>) &#8211; If it needs to read SNP data from a BED-formatted file, tells if it should count the number of A1
alleles (the PLINK standard) or the number of A2 alleles. False is the current default, but in the future the default will change to True.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">A phenotype dictionary with property &#8216;iid&#8217; listing the iids and property &#8216;vals&#8217; containing a nparray of PC values.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">logging</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">fastlmm.util</span> <span class="k">import</span> <span class="n">compute_auto_pcs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="s2">&quot;../feature_selection/examples/toydata&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">best_pcs</span> <span class="o">=</span> <span class="n">compute_auto_pcs</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="nb">int</span><span class="p">(</span><span class="n">best_pcs</span><span class="p">[</span><span class="s1">&#39;vals&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="nb">int</span><span class="p">(</span><span class="n">best_pcs</span><span class="p">[</span><span class="s1">&#39;vals&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">500 0</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="indices-and-tables">
<h2>Indices and Tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div></blockquote>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">fastlmm</span></code> API Documentation</a><ul>
<li><a class="reference internal" href="#single-snp"><code class="docutils literal"><span class="pre">single_snp</span></code></a></li>
<li><a class="reference internal" href="#single-snp-all-plus-select"><code class="docutils literal"><span class="pre">single_snp_all_plus_select</span></code></a></li>
<li><a class="reference internal" href="#single-snp-linreg"><code class="docutils literal"><span class="pre">single_snp_linreg</span></code></a></li>
<li><a class="reference internal" href="#single-snp-select"><code class="docutils literal"><span class="pre">single_snp_select</span></code></a></li>
<li><a class="reference internal" href="#epistasis"><code class="docutils literal"><span class="pre">epistasis</span></code></a></li>
<li><a class="reference internal" href="#snp-set"><code class="docutils literal"><span class="pre">snp_set</span></code></a></li>
<li><a class="reference internal" href="#fastlmm"><code class="docutils literal"><span class="pre">FastLMM</span></code></a></li>
<li><a class="reference internal" href="#linearregression"><code class="docutils literal"><span class="pre">LinearRegression</span></code></a></li>
<li><a class="reference internal" href="#heritability-spatial-correction"><code class="docutils literal"><span class="pre">heritability_spatial_correction</span></code></a></li>
<li><a class="reference internal" href="#compute-auto-pcs"><code class="docutils literal"><span class="pre">compute_auto_pcs</span></code></a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and Tables</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="#">fastlmm 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>